aten/src/ATen/native/Embedding.cpp

```diff
@@ -0,0 +1,181 @@
+
+  auto& sparse_type = grad.type().toBackend(grad.is_cuda() ? kSparseCUDA : kSparseCPU);
+  return sparse_type.sparse_coo_tensor(index, values);
+}
+
+Tensor embedding_backward_cpu(
+    const Tensor & grad_, const Tensor & indices, int64_t num_weights,
+    int64_t padding_idx, bool scale_grad_by_freq) {
+
+  auto indices_arg = TensorArg(indices, "indices", 2);
+  checkScalarType("embedding_backward", indices_arg, kLong);
+  checkContiguous("embedding_backward", indices_arg);
+
+  auto indices_data = indices.data<int64_t>();
+  int64_t numel = indices.numel();
+
+  std::unique_ptr<int64_t[]> counts;
+  if (scale_grad_by_freq) {
+    counts.reset(new int64_t[num_weights]);
+    for (int i = 0; i < numel; i++) {
+      counts[indices_data[i]] = 0;
+    }
+    for (int i = 0; i < numel; i++) {
+      counts[indices_data[i]]++;
+    }
+  }
+
+  auto grad = grad_.contiguous().view({numel, grad_.size(-1)});
+  auto grad_weight = grad_.type().zeros({num_weights, grad_.size(-1)});
+
+#ifdef _OPENMP
+  if (numel > 1000) {
+    // The strategy is to parallelize over sections of the vocabulary, so that
+    // thread 1 handles updates to gradWeight[0..nVocab/nThreads]. Every thread
+    // has to traverse the entire input, but the dominating factor is the axpy
+    // BLAS call.
+    #pragma omp parallel
+    {
+      int tid = omp_get_thread_num();
+      int nthreads = omp_get_num_threads();
+      int64_t start = tid * (num_weights/nthreads + 1);
+      int64_t end = start + (num_weights/nthreads + 1);
+      for (int64_t i = 0; i < numel; i++) {
+        if (indices_data[i] != padding_idx) {
+          int64_t k = indices_data[i] - TH_INDEX_BASE;
+          if (k >= start && k < end) {
+            double scale = 1.0;
+            if (scale_grad_by_freq) {
+              scale /= counts[k];
+            }
+            grad_weight[k].add_(grad[i], scale);
+          }
+        }
+      }
+    }
+    return grad_weight;
+  }
+#endif
+
+  for (int64_t i = 0; i < numel; i++) {
+    if (indices_data[i] != padding_idx) {
+      int64_t k = indices_data[i];
+      double scale = 1.0;
+      if (scale_grad_by_freq) {
+        scale /= counts[k];
+      }
+      grad_weight[k].add_(grad[i], scale);
+    }
+  }
+
+  return grad_weight;
+}
+
+Tensor & embedding_renorm_cpu_(
+    Tensor & self, const Tensor & indices, double max_norm, double norm_type) {
+  auto self_arg = TensorArg(self, "self", 1);
+  auto indices_arg = TensorArg(indices, "indices", 2);
+  checkContiguous("embedding_renorm_", self_arg);
+  checkDim("embedding_renorm_", self_arg, 2);
+  checkContiguous("embedding_renorm_", indices_arg);
+  checkScalarType("embedding_renorm_", indices_arg, kLong);
+
+  auto num_indices = indices.numel();
+  auto indices_data = std::unique_ptr<int64_t[]>(new int64_t[num_indices]);
+  std::memcpy(indices_data.get(), indices.data_ptr(), num_indices * sizeof(int64_t));
+  std::qsort(indices_data.get(), num_indices, sizeof(int64_t),
```

Comment:
Isn't `std::sort` more efficient than `std::qsort` (don't think the compiler can inline the check as a lambda, but `std::sort` is a template, so it's inlined for sure). Just use `std::vector` for indices, and take its `.data()`